afpacket/src/tokio_sockets.rs

```diff
@@ -20,11 +35,287 @@ impl AsyncBoundSocket {
+                Poll::Pending
+            }
+            Err(e) => Poll::Ready(Err(e)),
+        }
+    }
+
+    pub fn try_send(&mut self, frame: &[u8]) -> io::Result<usize> {
+        self.sock.get_mut().send(frame)
+    }
+
+    pub fn try_recv(&mut self, frame: &mut [u8]) -> io::Result<(usize, sockets::Addr)> {
+        self.sock.get_mut().recv(frame)
+    }
+
+    pub fn send<'a>(&'a mut self, frame: &'a [u8]) -> impl Future<Output = io::Result<usize>> + 'a {
+        SendFuture { sock: self, frame }
+    }
+
+    pub fn recv<'a>(
+        &'a mut self,
+        frame: &'a mut [u8],
+    ) -> impl Future<Output = io::Result<(usize, sockets::Addr)>> + 'a {
+        RecvFuture { sock: self, frame }
+    }
+}
+
+struct SendFuture<'a> {
+    sock: &'a mut AsyncBoundSocket,
+    frame: &'a [u8],
+}
+
+impl Unpin for SendFuture<'_> {}
+
+impl Future for SendFuture<'_> {
+    type Output = io::Result<usize>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let me = self.get_mut();
+        let sock = &mut me.sock;
+        let frame = &me.frame;
+        sock.poll_send(cx, &frame)
+    }
+}
+
+struct RecvFuture<'a> {
+    sock: &'a mut AsyncBoundSocket,
+    frame: &'a mut [u8],
+}
+
+impl Unpin for RecvFuture<'_> {}
+
+impl Future for RecvFuture<'_> {
+    type Output = io::Result<(usize, sockets::Addr)>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let me = self.get_mut();
+        let sock = &mut me.sock;
+        let mut frame = &mut me.frame;
+        sock.poll_recv(cx, &mut frame)
     }
+}
+
+pub struct RecvHalf {
+    inner: Arc<Inner>,
+}
+
+pub struct SendHalf {
+    inner: Arc<Inner>,
+}
+
+pub struct SendHalfLock<'a> {
+    guard: Guard<'a>,
+}
+
+impl RecvHalf {
+    pub fn is_pair_of(&self, other: &SendHalf) -> bool {
+        other.is_pair_of(self)
+    }
+
+    pub fn unsplit(self, tx: SendHalf) -> AsyncBoundSocket {
+        if self.is_pair_of(&tx) {
+            drop(tx);
+
+            let inner = Arc::try_unwrap(self.inner)
+                .ok()
+                .expect("Arc::try_unwrap failed");
+
+            inner.sock.into_inner()
+        } else {
+            panic!("Unrelated SendHalf passed to unsplit")
+        }
+    }
+
+    pub fn poll_can_rx(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        let mut inner = ready!(self.inner.poll_lock(cx));
+        inner.sock_pin().poll_can_rx(cx)
+    }
+
+    pub fn poll_recv(
+        &mut self,
+        cx: &mut Context<'_>,
+        frame: &mut [u8],
+    ) -> Poll<io::Result<(usize, sockets::Addr)>> {
+        let mut inner = ready!(self.inner.poll_lock(cx));
+        inner.sock_pin().poll_recv(cx, frame)
+    }
+
+    pub fn recv<'a>(
+        &'a mut self,
+        frame: &'a mut [u8],
+    ) -> impl Future<Output = io::Result<(usize, sockets::Addr)>> + 'a {
+        RecvHalfRecvFuture { half: self, frame }
+    }
+}
+
+struct RecvHalfRecvFuture<'a> {
```

Comment:
Let's get some doc comments on what all these structs are for. They won't be visible in the public facing documentation, but useful for maintenance.